--- a/xioinitialize.c	2019-04-04 10:59:55.000000000 +0200
+++ b/xioinitialize.c	2022-09-02 19:20:22.000000000 +0200
@@ -28,10 +28,15 @@
    assert(sizeof(uint32_t)==4);
 
    /* assertions regarding O_ flags - important for XIO_READABLE() etc. */
+#ifdef __GNU__
+   assert(O_RDONLY==1);
+   assert(O_WRONLY==2);
+   assert(O_RDWR==3);
+#else
    assert(O_RDONLY==0);
    assert(O_WRONLY==1);
    assert(O_RDWR==2);
-
+#endif
    assert(SHUT_RD==0);
    assert(SHUT_WR==1);
    assert(SHUT_RDWR==2);
